diff --git a/src/s2/s2builder_test.cc b/src/s2/s2builder_test.cc
index 0a01d32..7d8d522 100644
--- a/src/s2/s2builder_test.cc
+++ b/src/s2/s2builder_test.cc
@@ -15,6 +15,7 @@
 
 // Author: ericv@google.com (Eric Veach)
 
+#define __STDC_FORMAT_MACROS
 #include "s2/s2builder.h"
 
 #include <algorithm>
diff --git a/src/s2/s2builderutil_snap_functions_test.cc b/src/s2/s2builderutil_snap_functions_test.cc
index 83b651f..3a46a97 100644
--- a/src/s2/s2builderutil_snap_functions_test.cc
+++ b/src/s2/s2builderutil_snap_functions_test.cc
@@ -22,6 +22,7 @@
 // spherical case is slightly better or worse because of the spherical
 // distortion.
 
+#define __STDC_FORMAT_MACROS
 #include "s2/s2builderutil_snap_functions.h"
 
 #include <algorithm>
