From 682368288d1557f48e1acc4ca0a6eecbaa0463dc Mon Sep 17 00:00:00 2001
From: Bouwe Andela <b.andela@esciencecenter.nl>
Date: Fri, 6 Mar 2026 12:32:08 +0100
Subject: [PATCH 3/3] version

---
 pyproject.toml | 1 +
 1 file changed, 1 insertion(+)

diff --git a/pyproject.toml b/pyproject.toml
index 6a55b8c..9938160 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -56,3 +56,4 @@ namespaces = false
 [tool.setuptools.package-dir]
 "" = "src"
 
+[tool.setuptools_scm]
